packages/vm/src/index.ts

```diff
@@ -143,6 +166,7 @@ export default class VM extends AsyncEventEmitter {
   protected _opcodes: OpcodeList
   protected readonly _hardforkByBlockNumber: boolean
   protected readonly _hardforkByTD?: BNLike
+  protected readonly _customOpcodes?: CustomOpcode[]
```

Comment:
Totally a side topic here but just want to mention on the occasion since we are already so much in these StateManager design questions: my dream ("I have a dream", lol) for the breaking releases is still that we also manage to separate the EVM itself (so basically: everything in the `evm` folder) and all this outer execution stuff (so roughly: everything - at least - from `VM.runTx()` upwards).

This mangling together of the pure EVM execution logic (opcodes in, result out, basically 😋) and all this outer environment stuff with running txs, creating receipts, running the block, assiging block rewards in this one big `VM` object is one of the biggest design flaws we have. One can see this over and over again, e.g. this `customOpcodes` option would - if we would have that - also rather only be a thing to be passed to a clean (inner) `EVM` interface (in the sense of: API) and nothing to be babbled to this monster object with all the additional logic.

I guess this outer thing is something which rather would belong to the client then to the VM (we for sure doesn't want to integrate though) and historically likely was bapped on to the VM because there just wasn't a wrapping client or something and people needed this functionality.

Maybe we can come up with a separate package or so - `VMExecutionEnv` (I have totally no idea on a fitting name yet) - and then create a stable interface (again: API) to the EVM package.

This would lead to a lot of new use cases. The EVM package could be swapped out, extended by others, used in a standalone-context, etc.. I am also pretty sure that our overall VM architecture would greatly benefit by refactoring questions which come up along the way.

This would need some serious design thinking though, not sure if we would make it with this into this round of breaking releases. Maybe you also want to give this some thinking already. 🙂

P.S.: I know, comment is totally misplaced here. 😛

Reply:
I think the general point you are making here is that we should take a deep look at our internals in the VM to see if we should refactor there. I'm also getting the feeling that some things are "all over the place" and maybe a bit misplaced in some cases as well. We might do an internal refactor soon?

Reply:
I would like to do one along the breaking releases, with the optimal outcome that we can separate EVM and VM and have a clean (and somewhat refactored) API between them (what also comes into play here is this EVMC interface https://github.com/ethereum/evmc eventually).

Not sure if we'll make it though to full extend in this round, this is particularly hard.